src/topology/homotopy/homotopy_group.lean

```diff
@@ -5,141 +5,306 @@ Authors: Roberto Alvarez
+@[trans] lemma trans (H0 : homotopic f g) (H1 : homotopic g h) : homotopic f h := H0.trans H1
 
-lemma equiv : equivalence (@homotopic X _ n x) :=
+lemma equiv : equivalence (@homotopic X _ N x) :=
 ⟨homotopic.refl, λ _ _, homotopic.symm, λ _ _ _, homotopic.trans⟩
 
-instance setoid (n : ℕ) (x : X) : setoid (gen_loop n x) := ⟨homotopic, equiv⟩
+instance setoid (N) (x : X) : setoid (gen_loop N x) := ⟨homotopic, equiv⟩
 
 end
 end homotopic
 
+section
+
+variable [decidable_eq N]
+
+/-- Loop from a generalized loop by currying $I^N → X$ into $I → (I^{N\setminus\{j\}} → X)$. -/
+@[simps] def to_loop (i : N) : gen_loop N x → Ω (gen_loop {j // j ≠ i} x) const := λ p,
+{ to_fun := λ t, ⟨(p.val.comp (cube.insert_at i).to_continuous_map).curry t,
+    λ y yH, p.property (cube.insert_at i (t, y)) (cube.insert_at_boundary i $ or.inr yH)⟩,
+  source' := by { ext t, refine p.property (cube.insert_at i (0, t)) ⟨i, or.inl _⟩, simp },
+  target' := by { ext t, refine p.property (cube.insert_at i (1, t)) ⟨i, or.inr _⟩, simp } }
+
+/-- Generalized loop from a loop by uncurrying $I → (I^{N\setminus\{j\}} → X)$ into $I^N → X$. -/
+@[simps] def from_loop (i : N) (p : Ω (gen_loop {j // j ≠ i} x) const) : gen_loop N x :=
+⟨(⟨λ t, (p t).1, by continuity⟩ : C(I, C(cube _, X))).uncurry.comp
+  (cube.split_at i).to_continuous_map,
+begin
+  rintros y ⟨j, Hj⟩,
+  simp only [subtype.val_eq_coe, continuous_map.comp_apply, to_continuous_map_apply,
+    fun_split_at_apply, continuous_map.uncurry_apply, continuous_map.coe_mk,
+    function.uncurry_apply_pair],
+  obtain rfl | Hne := eq_or_ne j i,
+  { cases Hj; rw Hj; simp only [p.source, p.target]; convert const_apply },
+  { exact gen_loop.boundary _ _ ⟨⟨j, Hne⟩, Hj⟩ },
+end⟩
+
+lemma to_from (i : N) (p : Ω (gen_loop {j // j ≠ i} x) const) : to_loop i (from_loop i p) = p :=
+begin
+  simp_rw [to_loop, from_loop, continuous_map.comp_assoc, to_continuous_map_as_coe,
+    to_continuous_map_comp_symm, continuous_map.comp_id], ext, refl,
+end
+
+/-- The `n+1`-dimensional loops are in bijection with the loops in the space of
+  `n`-dimensional loops with base point `const`.
+  We allow an arbitrary indexing type `N` in place of `fin n` here. -/
+@[simps] def loop_equiv (i : N) : gen_loop N x ≃ Ω (gen_loop {j // j ≠ i} x) const :=
+{ to_fun := to_loop i,
+  inv_fun := from_loop i,
+  left_inv := λ p, by { ext, exact congr_arg p (equiv.apply_symm_apply _ _) },
+  right_inv := to_from i }
+
+lemma to_loop_apply (i : N) {p : gen_loop N x} {t} {tn} :
+  to_loop i p t tn = p (cube.insert_at i ⟨t, tn⟩) := rfl
+
+lemma from_loop_apply (i : N) {p : Ω (gen_loop {j // j ≠ i} x) const} {t : cube N} :
+  from_loop i p t = p (t i) (cube.split_at i t).snd := rfl
+
+end
+
+section
+
+/-- The inclusion from the space of generalized loops to the space of all continuous functions
+  (not necessarily constant on the boundary), as a continuous map. -/
+@[reducible] def c_coe : C(gen_loop N x, C(cube N, X)) := ⟨λ p, p.val, continuous_induced_dom⟩
```

Comment:
We're missing a more general definition here,
```lean
/-- `subtype.val` as a continuous function -/
def continuous_map.subtype {α} (p : α → Prop) [topological_space α] : C(subtype p, α) :=
{ to_fun := coe, continuous_to_fun := continuous_induced_dom }
```
which belongs in an earlier file.

Reply:
Probably not worth modifying another ported file, given that `c_coe` is [easily inlined](https://github.com/leanprover-community/mathlib/pull/15681/commits/c953156ebaed677e80cd14d3ad0c026349e259dd#diff-8d28e1c85b22df3c99f61de0a89bc1e7ba35a56eae6db74bcdec15750d007e50L250-R241).